Objects/longobject.c

```diff
@@ -4502,37 +4507,71 @@ long_rshift1(PyLongObject *a, Py_ssize_t wordshift, digit remshift)
         return _PyLong_FromSTwoDigits(x);
     }
 
-    if (Py_SIZE(a) < 0) {
-        /* Right shifting negative numbers is harder */
-        PyLongObject *a1, *a2;
-        a1 = (PyLongObject *) long_invert(a);
-        if (a1 == NULL)
-            return NULL;
-        a2 = (PyLongObject *) long_rshift1(a1, wordshift, remshift);
-        Py_DECREF(a1);
-        if (a2 == NULL)
-            return NULL;
-        z = (PyLongObject *) long_invert(a2);
-        Py_DECREF(a2);
+    a_negative = Py_SIZE(a) < 0;
+    size_a = Py_ABS(Py_SIZE(a));
+
+    if (a_negative) {
+        /* For 'a', adjust so that 0 < remshift <= PyLong_SHIFT. This ensures
+           that 'newsize' is computed correctly below. */
+        if (remshift == 0) {
+            remshift = PyLong_SHIFT;
+            --wordshift;
+            if (wordshift < 0) {
+                /* Can only happen if the original shift was 0. */
```

Comment:
Would it make more sense to check for `wordshift == 0` before the decrement?
It would be clearer (to me, at least).